contents/javascript/Javascript-promises.md

```diff
@@ -89,58 +87,117 @@ These situations correspond to the 3 states of Promises in Javascript.
 
 ```javascript
+function executeWithDelay(val, callback) {
+    // return val after a short wait
+    setTimeout(function(){
+        callback(val);
+    }, 1000);
+}
+
+/** Delayed function calls **/
+function getData(callback) {
+    executeWithDelay('some random data', callback);
+}
+
+function filterData(data, callback) {
+    executeWithDelay(data.split(' '), callback);
+}
+
+function main() {
+    getData(function(data){
+        filterData(function(filtered){
+            // will print array of splitted text
+            console.log(filtered);
+        })
+    });
+}
+
+main();
+```
+
+Now we will rewrite all these using promises. We will be using the same function and variable names, to show how exactly promise compare to callbacks. 
 
-function update() {
-    return new Promise((resolve, reject) => {
-        var x = "I have not been updated";
-        setTimeout(function() {
-            x = "I have been updated";
-            resolve(x);
+```javascript
+function executeWithDelay(val) {
+    // return val after a short wait
+    return new Promise(resolve => {
+        setTimeout(function(){
+            resolve(val);
         }, 1000);
     });
 }
 
-update().then(result => console.log(result));
+/** note these functions now return a promise **/
+function getdata() {
+    return executeWithDelay('some random data');
+}
+
+function filterdata(data, callback) {
+    return executeWithDelay(data.split(' '));
+}
+
+function main() {
+    return getData()
+        .then(data => filterData(data))
+        .then(filtered => console.log(filtered));
+}
 
+main();
 ```
 
-The first step of creating a Promise involves invoking the `Promise` constructor which takes in just 1 parameter known as the executor function. This is the function that will be executed by the Promise. This function has two parameters: resolve, reject. In our example, we only use the resolve function to indicate that our promise has executed successfully with the result `x`. The reject function may be used to indicate that some error has occured during execution. 
+Using Promises keep the code clean and easy to read, which makes it easy to learn what the code does, instead of having to trace through the code written with callbacks.
 
-After defining the function, we call it, and we call another function `.then()` upon the result. `.then()` is used to consume the result from a fulfilled Promise and act on it. In this case, we are using it to log our result. More importantly, it lets us chain promises to avoid the deep nesting known as callback hell.
+### The imperative style of promises
 
-Here's the callback hell code rewritten with Promises.
+The 'promise style', which uses `.then()` to pass data from one function to the next is comparable to functional programming. The original promise is passed from one `.then()` to the other, and with each `.then()`, a new promise is returned for the next `.then()` to work on.
```

Comment:
`comparable to functional programming` -> `is typical to _functional programming_`